### PR TITLE
prometheus: add package

### PR DESCRIPTION
**Pull request by aparcar**
Maintainer: me
Compile tested: x86/64, aarch64_cortex-a53
Run tested: x86/64, aarch64_cortex-a53

Description:

[From official GitHub page:](https://github.com/prometheus/prometheus/)

---

Prometheus, a [Cloud Native Computing Foundation](https://cncf.io/) project, is a systems and service monitoring system. It collects metrics
from configured targets at given intervals, evaluates rule expressions,
displays the results, and can trigger alerts if some condition is observed
to be true.

Prometheus' main distinguishing features as compared to other monitoring systems are:

- a **multi-dimensional** data model (timeseries defined by metric name and set of key/value dimensions)
- a **flexible query language** to leverage this dimensionality
- no dependency on distributed storage; **single server nodes are autonomous**
- timeseries collection happens via a **pull model** over HTTP
- **pushing timeseries** is supported via an intermediary gateway
- targets are discovered via **service discovery** or **static configuration**
- multiple modes of **graphing and dashboarding support**
- support for hierarchical and horizontal **federation**

---

As prometheus is written in Go it requires the golang package to be
installed. Currently is golang still a pending pull request[1].

[1] https://github.com/openwrt/packages/pull/5780

Signed-off-by: Paul Spooren <spooren@informatik.uni-leipzig.de>

### CODE REVIEW

**Comment by aparcar** [comment]
Dear @jefferyto, if you find the time please check if the Makefile looks good. I don't know how to separate the package into single libraries - nor if that's actually required.

File: utils/prometheus/Makefile [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#
+# Copyright (C) 2018 Jeffery To
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: Can replace my name with yours :smiley:

File: utils/prometheus/Makefile [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=prometheus
+PKG_VERSION:=2.2.1
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/prometheus/prometheus.git
+PKG_SOURCE_VERSION:=f8dcf9b272c9ab1a6148ccc22c4de6d4403c601d
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: Would be great if you can include a [`PKG_MIRROR_HASH`](https://openwrt.org/docs/guide-developer/packages#use_source_repository) (and preferable `PKG_SOURCE_DATE` also).

File: utils/prometheus/Makefile [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=LICENSE
+PKG_MAINTAINER:=Paul Spooren <spooren@informatik.uni-leipzig.de>
+
+PKG_BUILD_DEPENDS:=golang/host
+PKG_BUILD_PARALLEL:=1
+PKG_USE_MIPS16:=0
+
+GO_PKG:=github.com/prometheus/prometheus/
+GO_PKG_BUILD_PKG:=github.com/prometheus/prometheus/cmd/prometheus/
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: This only builds `prometheus` and not `promtool`, is that the intended behaviour? Would `promtool` be useful on the target device? (FYI I've never used Prometheus so I don't know what `promtool` does.)

File: utils/prometheus/Makefile [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+define Package/prometheus/Default
+  TITLE:=Monitoring system & time series database
+  URL:=http://prometheus.io
+  DEPENDS:=$(GO_ARCH_DEPENDS)
+endef
+
+define Package/prometheus
+$(call Package/prometheus/Default)
+  TITLE+= (source files)
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: I believe you're building a binary-only package, so the "(source files)" is unnecessary.

File: utils/prometheus/Makefile [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  PKGARCH:=all
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: This would not work for a binary package (can just remove this line).

File: utils/prometheus/Makefile [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  CATEGORY:=Utilities
+endef
+
+define Package/prometheus/description
+Prometheus, a Cloud Native Computing Foundation project, is a systems and
+service monitoring system. It collects metrics from configured targets at given
+intervals, evaluates rule expressions, displays the results, and can trigger
+alerts if some condition is observed to be true.
+
+https://github.com/prometheus/prometheus
```
Comment (jefferyto) [inline utils/prometheus/Makefile]: Including the repo URL here isn't necessary since the homepage URL is already included in the package details above.

**Comment by jefferyto** [comment]
Thanks for submitting this! On separating out build dependencies, personally I would prefer it if dependencies were packaged separately, but from [Debian's dependency list](https://packages.debian.org/source/sid/prometheus) I can understand it looks like a lot of work :laughing:

So I wouldn't say it's necessary for now, but I would look to split out dependencies when more than one application depends on the same library/framework/etc.

(As for "how" to separate libraries out, after adding a source package for the library and including that source package as a build dependency, we would need to delete the library from the application's `vendor` directory. I may add code to make that easier later on.)

File: utils/prometheus/Makefile (hunk 2 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: Added

File: utils/prometheus/Makefile (hunk 3 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: `promtool` is quite big and so I though of rather validation the configurations locally and only install the basic `prometheus` daemon on the node.

File: utils/prometheus/Makefile (hunk 4 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: Removed

File: utils/prometheus/Makefile (hunk 5 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: Removed

File: utils/prometheus/Makefile (hunk 6 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: Removed

File: utils/prometheus/Makefile (hunk 1 above)
Comment (aparcar) [reply utils/prometheus/Makefile]: Oops, removed. I don't intent to have any copyright on this...

**Comment by aparcar** [comment]
@jefferyto now that golang was merged into master, do you think something has to be changed to merge this?

**Comment by egon0** [comment]
i hope this gets merged until 18.06 is finally freezed. would love to see it in the next official release so i could stop working with openwrt-snapshots :)

**Comment by aparcar** [comment]
Updated Prometheus to 2.3.0 and added UCI config options for storage and tsdb path.

**Comment by aparcar** [comment]
would it make sense to add `procd_set_param user prometheus` to run as non-root?

**Comment by aparcar** [comment]
updated to prometheus 2.3.1 and alertmanager 0.15
use codeload in both cases

anything stops this from merging?

**Comment by aparcar** [comment]
Update to Prometheus 2.3.2 and removed the `prometheus-alertmanager` to make the PR smaller.

It's tested and works, can someone please merge this? I need this for a report and it would be very nice to refer to OpenWrt snapshots rather than providing binary packages myself.

**Pull request merged**